Switch to MonadIO

lehins opened the pull request:
Switches all exposed actions to MonadIO, which makes it easy to use with various monad transformers and plethora of other libraries. Slight theoretical breaking change, but benign in practice, hence major version bump.

~This PR depends on #4 ~

lehins committed:
Switch all exposed `IO` actions to `MonadIO`

GianlucaGuarini commented:
Could you please add more info about this change? Subclassing the IO Monad makes the code harder to read and increases its complexity. This project is meant to be used as standalone cli so I don't get the point about the polymorphic libs support. Thanks

lehins commented:
I used it as a library within [`RIO`](https://www.stackage.org/haddock/nightly-2023-08-17/rio-0.1.22.0/RIO.html#t:RIO) monad, which is great for building applications. However, anytime you need to use an actual `IO` within any other monad that is based on IO you are forced to use `liftIO`, which is a bit of an inconvenience.

It doesn't have to be RIO, if my application uses a stack of transformers and I'd like to interact with a user within any of my functions, problem is the same. Switching it to MonadIO simplifies things.

Here is a concrete example:

```haskell
writeToFileWithConfirmation
  :: (HasLogFunc env, MonadReader env m, MonadIO m)
  => FilePath
  -> Text
  -> m ()
writeToFileWithConfirmation filePath contents = do
  reportAlreadyExists <- doesFileExist filePath
  proceed <-
    if reportAlreadyExists
      then
        confirmWithDefault
          ("File with name: " <> show filePath <> " already exists! Overwrite?")
          False
      else pure True
  when proceed $ do
    writeBinaryFile filePath $ encodeUtf8 contents
    logInfo $ "Wrote Report to: " <> display (T.pack filePath)
```

Note that I didn't have to do any lifting.

The important part all of the actions that were switched to MonadIO will still work just fine in `IO`

GianlucaGuarini commented:
After a bit of research I still don't get why the `MonadIO` subclassing should add better compatibility with other libs. 

[This article](https://chrispenner.ca/posts/monadio-considered-harmful) recommends splitting the monads into smaller semantic components that can be combined into transformers to avoid the `MondadIO` overhead.

Could you please add more examples (before and after the patch) to show how this change might improve the use of this lib? Thank you

lehins commented:
This is how the example above would look like if I was using functions that are not using `MonadIO`:

* [`doesFileExist`](https://hackage.haskell.org/package/directory-1.3.8.1/docs/System-Directory.html#v:doesFileExist) vs [`doesFileExist`](https://hackage.haskell.org/package/unliftio-0.2.25.0/docs/UnliftIO-Directory.html#v:doesFileExist)
* [`writeBinaryFile`](https://hackage.haskell.org/package/unliftio-0.2.25.0/docs/UnliftIO-IO-File.html#v:writeBinaryFile) vs [`writeFile`](https://hackage.haskell.org/package/bytestring-0.12.0.2/docs/Data-ByteString.html#v:writeFile)
* `confirmWithDefault` in released version vs in this PR
```haskell
writeToFileWithConfirmation
  :: (HasLogFunc env, MonadReader env m, MonadIO m)
  => FilePath
  -> Text
  -> m ()
writeToFileWithConfirmation filePath contents = do
  reportAlreadyExists <- liftIO $ doesFileExist filePath
  proceed <-
    if reportAlreadyExists
      then
        liftIO $ confirmWithDefault
          ("File with name: " <> show filePath <> " already exists! Overwrite?")
          False
      else pure True
  when proceed $ do
    liftIO $ writeBinaryFile filePath $ encodeUtf8 contents
    logInfo $ "Wrote Report to: " <> display (T.pack filePath)
```

Note that you can't restrict `m` to `IO` in the above function because logging imposes: `MonadReader`, so you have to call `liftIO`.

For current examples we are only talking about `MonadIO`, but for more complex examples there is potentially a need for `MonadUnliftIO` too. It is a common pattern in Haskell ecosystem, so much so that `MonadIO` is provided by `base`.

That being said. Feel free to close this PR if you don't see the benefit. It does not impact me in any way, I doubt that I will be an avid user of this library :) I just suggested this change, because it is good practice.

lehins commented:
FYI, the article "MonadIO Considered Harmful" is just an opinion of one person, don't take it too seriously :wink:

GianlucaGuarini commented on src/FortyTwo/Prompts/Confirm.hs:
```diff
@@ -14,18 +15,18 @@ normalizeString :: String -> String
 -- | Ask a confirm falling back to a default value if no answer will be provided
-confirmWithDefault :: String -> Bool -> IO Bool
+confirmWithDefault :: MonadIO m => String -> Bool -> m Bool
 confirmWithDefault question defaultAnswer = do
```
Can't we wrap the whole function in a `liftIO` call?

lehins replied:
Yeah, it makes no difference, really. I've wrapped the whole function.

GianlucaGuarini reviewed (approved):
Thank you @lehins . After a bit of research I understood the nature of this CR.
I have learnt a lot with it. I have just a small question but I am ready to merge it and bump a major release

lehins commented:
@GianlucaGuarini I am really glad you found this PR useful. The whole concept of MonadIO is omnipresent in Haskell ecosystem. So, it is certainly a good thing to be familiar with.

I've addressed your comment too.
